File: q1488/q1488.java
class Solution {
    public int[] avoidFlood(int[] rains) {
        int[] res = new int[rains.length];
        Arrays.fill(res, 1);
        TreeSet<Integer> prevDryDays = new TreeSet<Integer>();
        Map<Integer, Integer> lakeToPrevFullDays = new HashMap<>();
        for (int day = 0; day < rains.length; day++) {
            int lake = rains[day];
            // if it is a dry day
            if (lake == 0) prevDryDays.add(day);
            else {
                res[day] = -1;
                // if this lake is already full
                if (lakeToPrevFullDays.containsKey(lake)) {
                    int prevRainDay = lakeToPrevFullDays.get(lake);
                    Integer dryDay = prevDryDays.ceiling(prevRainDay);
                    if (dryDay == null) return new int[0];
                    res[dryDay] = lake;
                    prevDryDays.remove(dryDay);
                }
                lakeToPrevFullDays.put(lake, day);
            }
        }
        return res;
    }
}

File: q1488/q1488.py
class Solution:
    def avoidFlood(self, rains: List[int]) -> List[int]:
        res = [1] * len(rains)
        sorted_dry_days = SortedList()
        lake_to_day = dict()
        for day, lake in enumerate(rains): 
            if lake == 0:
                sorted_dry_days.add(day)
            else: 
                res[day] = -1
                if lake in lake_to_day:
                    rainy_day = lake_to_day[lake]
                    # search for the first day before the rainy days
                    dry_day_idx = sorted_dry_days.bisect(rainy_day)
                    if (dry_day_idx == len(sorted_dry_days)): return []
                    dry_day = sorted_dry_days[dry_day_idx]
                    res[dry_day] = lake
                    sorted_dry_days.discard(dry_day)
                lake_to_day[lake] = day
        return res
               
